User (family) authentication middleware and create user profile API

RohithKasar opened the pull request:
### Administrative Info
Monday Board ID: 853642866
Make sure your branch name conforms to: `<feature/staging/hotfix/...>/[username]/[Monday Item ID]-[3-4 word description separated by dashes]`. Otherwise, please rename your branch and create a new PR.

### Changes
What changes did you make?
Created user model, can now create users given a firebase uid
Wrote authentication middleware that ensures user has valid token before proceeding

### Testing
How did you confirm your changes worked? 
Attempted to create users with no token, invalid token, valid token, and duplicate users and asserted the appropriate status codes were returning
### Confirmation of Change 
Upload a screenshot, if possible. Otherwise, please provide instructions on how to see the change.

run make test in backend

dacruz21 commented:
Neither of you have your .gitconfig set up correctly, can you double-check the name and email address and make sure the email matches the one you use on GitHub?

![image](https://user-images.githubusercontent.com/36964941/109759835-8a4aac00-7ba2-11eb-8d16-f75b91118848.png)

dacruz21 commented on backend/internal/auth/auth.go:
```diff
@@ -0,0 +1,9 @@
+	"context"
+)
+
+type Authenticator interface {
+	VerifyToken(ctx context.Context, token string) (string, bool)
```
Documentation on this interface and function would be useful.

dacruz21 commented on backend/internal/auth/firebase.go:
```diff
@@ -0,0 +1,18 @@
+	client auth.Client
+}
+
+func (a *FirebaseAuthenticator) VerifyToken(ctx context.Context, authHeader string) (string, bool) {
```
Documentation on this implementation would be helpful (mention that it verifies the JWT is signed and is valid for the Firebase project)

dacruz21 replied:
Also, this function does *not* take the entire Authorization header, it just takes the token. Let's rename `authHeader` to `token`

dacruz21 commented on backend/internal/auth/mock.go:
```diff
@@ -0,0 +1,16 @@
+
+type MockAuthenticator struct{}
+
+func (a MockAuthenticator) VerifyToken(ctx context.Context, token string) (string, bool) {
```
Document this with the structure of test tokens (that they are of the form `test-token<user ID>`)

dacruz21 commented on backend/internal/controllers/main_test.go:
```diff
@@ -0,0 +1,63 @@
+	_, _ = conn.Exec("TRUNCATE books")
+	_, _ = conn.Exec("TRUNCATE book_contents")
+	_, _ = conn.Exec("TRUNCATE users")
+
+	// Populate database
+	conn.Exec("INSERT INTO books (id, title, author) values ('c_id', 'c','c1');")
+	conn.Exec("INSERT INTO books (id, title, author) values ('a_id', 'a','a1');")
+	conn.Exec("INSERT INTO books (id, title, author) values ('b_id', 'b','b1');")
+	conn.Exec("INSERT INTO books (id, title, author) VALUES ('catcher'," +
+		"'catcher in the rye', 'a');")
+
+	conn.Exec("INSERT INTO book_contents (id, lang, read_video, read_body, " +
+		"explore_video, explore_body, learn_video, learn_body) VALUES " +
+		"('a_id', 'en', 'a_en_rv', 'a_en_rb', 'a_en_ev', 'a_en_eb', " +
+		"'a_en_lv', 'a_en_lb')")
+
+	conn.Exec("INSERT INTO book_contents (id, lang, read_video, read_body, " +
+		"explore_video, explore_body, learn_video, learn_body) VALUES " +
+		"('b_id', 'en', 'b_en_rv', 'b_en_rb', 'b_en_ev', 'b_en_eb', " +
+		"'b_en_lv', 'b_en_lb')")
+
+	conn.Exec("INSERT INTO book_contents (id, lang, read_video, read_body, " +
+		"explore_video, explore_body, learn_video, learn_body) VALUES " +
+		"('c_id', 'en', 'c_en_rv', 'c_en_rb', 'c_en_ev', 'c_en_eb', " +
+		"'c_en_lv', 'c_en_lb')")
+
+	conn.Exec("INSERT INTO book_contents (id, lang, read_video, read_body, " +
+		"explore_video, explore_body, learn_video, learn_body) VALUES " +
+		"('a_id', 'es', 'a_es_rv', 'a_es_rb', 'a_es_ev', 'a_es_eb', " +
+		"'a_es_lv', 'a_es_lb')")
+
+	conn.Exec("INSERT INTO book_contents (id, lang, read_video, read_body, " +
+		"explore_video, explore_body, learn_video, learn_body) VALUES " +
+		"('catcher', 'en', 'catcher_rv', 'catcher_rb', 'catcher_ev', " +
+		" 'catcher_eb', 'catcher_lv', 'catcher_lb')")
+
+	conn.Exec("INSERT INTO book_contents (id, lang, read_video, read_body, " +
+		"explore_video, explore_body, learn_video, learn_body) VALUES " +
+		"('catcher', 'es', 'catcher_es_rv', 'catcher_es_rb', 'catcher_es_ev', " +
+		" 'catcher_es_eb', 'catcher_es_lv', 'catcher_es_lb')")
```
I think we'll want to move all of this test data back into the test files they are associated with, but I'm not sure how best to accomplish that. This is fine for now, I'll experiment in a separate PR

dacruz21 commented on backend/internal/controllers/middleware/requireAuth.go:
```diff
@@ -0,0 +1,41 @@
+	"github.com/TritonSE/words-alive/internal/auth"
+)
+
+// HTTP middleware setting a value on the request context
```
This comment is so vague, it should definitely mention that it checks the Authorization header on each request against the Authenticator it is given.

dacruz21 commented on backend/internal/controllers/middleware/requireAuth.go:
```diff
@@ -0,0 +1,41 @@
+// HTTP middleware setting a value on the request context
+func RequireAuth(authenticator auth.Authenticator) func(http.Handler) http.Handler {
+
+	// middleware returned on calling RequireAuth
```
IMO this comment is unnecessary, we all know what a `return` statement does :) Was there something else you were trying to say here?

dacruz21 commented on backend/internal/controllers/middleware/requireAuth.go:
```diff
@@ -0,0 +1,41 @@
+			// verify the token of the user
+			authHeader := req.Header.Get("Authorization")
+			if authHeader == "" {
+				// kill request
```
unnecessary I think, `writeResponse` is a pretty common, self-explaining pattern in our codebase

dacruz21 commented on backend/internal/controllers/middleware/requireAuth.go:
```diff
@@ -0,0 +1,41 @@
+			tokenString := strings.TrimPrefix(authHeader, "Bearer ")
+			uid, ok := authenticator.VerifyToken(req.Context(), tokenString)
+			if !ok {
+				// something went wrong
```
This is also vague, and I think unnecessary

dacruz21 commented on backend/internal/controllers/middleware/utils.go:
```diff
@@ -0,0 +1,12 @@
+
+// Sends data and status code for http
+func writeResponse(rw http.ResponseWriter, statusCode int, body interface{}) {
+	rw.WriteHeader(statusCode)
```
You need to set the Content-Type header here, just like in `controllers.writeResponse`.

dacruz21 commented on backend/internal/controllers/user.go:
```diff
@@ -0,0 +1,59 @@
+
+type UserController struct {
+	// Hold a reference to the database so that all of the functions below can use it
+	// Keeps an authenticator to authenticate token
```
Not anymore it doesn't, delete this? :)

dacruz21 replied:
And actually I think the comment above is unnecessary too, this is a common pattern that's pretty self-documenting

dacruz21 commented on backend/internal/controllers/user.go:
```diff
@@ -0,0 +1,59 @@
+	Users database.UserDatabase
+}
+
+func (c *UserController) CreateUser(rw http.ResponseWriter, req *http.Request) {
```
This handler needs a comment though

dacruz21 commented on backend/internal/controllers/user.go:
```diff
@@ -0,0 +1,59 @@
+		return
+	}
+
+	_, err := c.Users.FetchUserByEmail(req.Context(), user.Email)
+	if err != nil {
+		// we have created a user with this email before
+		writeResponse(rw, http.StatusBadRequest, "user with email already exists")
+		return
+	}
```
I don't think this is correct. Right now this will say "user with email already exists" if we can't run the query at all (database is down, SQL syntax error, etc.).

We should keep this `if err != nil` block and throw a 500 in it, but what's more important is checking if the returned user (the first parameter that you've ignored with a `_`) is non-nil, i.e. a user already exists with that email.

dacruz21 commented on backend/internal/controllers/user.go:
```diff
@@ -0,0 +1,59 @@
+
+	id, err := c.Users.CreateUser(req.Context(), user)
+	if err != nil {
+		writeResponse(rw, http.StatusInternalServerError, "error creating user")
```
5xx errors should not have useful messages, can you change this back to `"error"`?

dacruz21 commented on backend/internal/controllers/user.go:
```diff
@@ -0,0 +1,59 @@
+
+	newUser, err := c.Users.FetchUserByID(req.Context(), id)
+	if err != nil {
+		writeResponse(rw, http.StatusInternalServerError, "error fetching user")
```
Same here

dacruz21 commented on backend/internal/database/users.go:
```diff
@@ -0,0 +1,63 @@
+
+import (
+	"context"
+	// 	"fmt"
```
Remove this please?

dacruz21 commented on backend/internal/models/user.go:
```diff
@@ -0,0 +1,12 @@
+package models
+
+/*
+User model for PostGres backend
+*/
```
1. C-style comment syntax isn't necessary for just one line
2. This model isn't specific to Postgres at all, it's intended to be the standard representation of a user across the entire backend.
3. The G in Postgres is lowercase :)

It would be much more useful to explain that this represents a end user of the app (a family, not an admin), and that the ID is the UID of an account in Firebase Auth

dacruz21 commented on backend/internal/testutils/testutils.go:
```diff
@@ -3,22 +3,51 @@ package testutils
+	res, err := http.DefaultClient.Do(req)
 	require.NoError(t, err)
+
+	require.Equal(t, expectedStatusCode, res.StatusCode)
+
+	defer res.Body.Close()
+	require.Equal(t, expectedStatusCode, res.StatusCode)
```
Why are you checking this twice? Also you should defer immediately after require.NoError()

typokign reviewed (changes requested):
Tests are really good, thanks for those. This could be better documented though, see comments

dacruz21 commented on backend/internal/controllers/router.go:
```diff
@@ -41,5 +45,8 @@ func GetRouter() chi.Router {
 		r.Get("/", bookController.GetBookList)
 	})
 
+	authenticator := auth.MockAuthenticator{}
```
Let's make GetRouter() take an authenticator as a parameter, and let's define this mock authenticator in main.go, with a big // TODO that we need to set up the Firebase project :)

dhrvsood committed:
User (family) authentication middleware and create user profile API

Create mock authenticator to test to simplify
authentication testing using defined token.

Create middleware folder and write requireAuth.go
to validate Authorization header and handle
authentication.

Write create user profile API in user.go to
handle errors and output correct HTTP code
to response.

Write tests in user_test.go to test for failure
and success of /users API route.

dacruz21 commented on .gitignore:
```diff
@@ -1,2 +1,4 @@
 .idea/
 .DS_Store
+*.env
+commit.txt
```
not sure how this got in here, remove please? otherwise LGTM!

dhrvsood committed:
Remove commit.txt from gitignore

typokign reviewed (approved):
![lgtm](https://i.imgur.com/xvWUf09.png)

k6han commented on backend/internal/auth/mock.go:
```diff
@@ -0,0 +1,17 @@
+func (a MockAuthenticator) VerifyToken(ctx context.Context, token string) (string, bool) {
+
+	if strings.HasPrefix(token, "test-token-") {
+		return strings.TrimPrefix(token, "test-token-"), true
```
Incredibly minor but this could be modified so that the structure is
if (problem)
      return false

return true

In the same way that the Firebase VerifyToken is written. Most of our other code also has the structure of checking for problems in if statements, and only returning true at the end

k6han commented:
LGTM. Nothing important needs any changes
